File: backend/app/conductor/schemas/agent.py
import uuid
from datetime import datetime
from enum import Enum
from typing import Any, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, field_serializer, model_validator


class EngineKind(str, Enum):
    CLAUDE = "claude"
    CODEX = "codex"


class ModelConfig(BaseModel):
    id: str
    speed: str = "standard"


class PermissionPolicy(BaseModel):
    type: Literal["always_allow", "always_ask"] = "always_allow"

    def to_mode_str(self) -> str:
        if self.type == "always_allow":
            return "bypassPermissions"
        return "default"


class ToolDefaultConfig(BaseModel):
    permission_policy: PermissionPolicy = Field(default_factory=PermissionPolicy)
    enabled: bool = True


class ToolItemConfig(BaseModel):
    name: str
    enabled: bool = True
    permission_policy: Optional[PermissionPolicy] = None


class AgentToolsetTool(BaseModel):
    type: Literal["agent_toolset_20260401"] = "agent_toolset_20260401"
    default_config: Optional[ToolDefaultConfig] = None
    configs: list[ToolItemConfig] = Field(default_factory=list)


class McpToolsetTool(BaseModel):
    type: Literal["mcp_toolset"] = "mcp_toolset"
    mcp_server_name: str
    default_config: Optional[ToolDefaultConfig] = None
    configs: list[ToolItemConfig] = Field(default_factory=list)


class CustomTool(BaseModel):
    type: Literal["custom"] = "custom"
    name: str
    description: str
    input_schema: dict[str, Any] = Field(default_factory=dict)


AgentTool = Union[AgentToolsetTool, McpToolsetTool, CustomTool]


class McpServerConfig(BaseModel):
    type: str = "url"
    name: str
    url: str


class PackedItem(BaseModel):
    name: str
    tar_gz_b64: str


class CreateAgentRequest(BaseModel):
    name: str
    engine_kind: EngineKind = EngineKind.CLAUDE
    model: Union[str, ModelConfig, None] = None
    system: Optional[str] = Field(default=None, alias="system_prompt")
    description: Optional[str] = None
    metadata: dict[str, str] = Field(default_factory=dict)
    env: dict[str, str] = Field(default_factory=dict)
    mcp_servers: list[McpServerConfig] = Field(default_factory=list)
    skills: list[PackedItem] = Field(default_factory=list)
    agents: list[PackedItem] = Field(default_factory=list)
    commands: list[PackedItem] = Field(default_factory=list)
    tools: list[AgentTool] = Field(default_factory=list)
    multiagent: Optional[dict[str, Any]] = None
    environment_ref: Optional[str] = None
    secret_ref: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="before")
    @classmethod
    def normalize_model(cls, data: Any) -> Any:
        if isinstance(data, dict):
            m = data.get("model")
            if isinstance(m, str):
                data["model"] = {"id": m, "speed": "standard"}
        return data


class UpdateAgentRequest(BaseModel):
    version: int
    name: Optional[str] = None
    engine_kind: Optional[EngineKind] = None
    model: Union[str, ModelConfig, None] = None
    system: Optional[str] = Field(default=None, alias="system_prompt")
    description: Optional[str] = None
    metadata: Optional[dict[str, str]] = None
    env: Optional[dict[str, str]] = None
    mcp_servers: Optional[list[McpServerConfig]] = None
    skills: Optional[list[PackedItem]] = None
    agents: Optional[list[PackedItem]] = None
    commands: Optional[list[PackedItem]] = None
    tools: Optional[list[AgentTool]] = None
    multiagent: Optional[dict[str, Any]] = None
    environment_ref: Optional[str] = None
    secret_ref: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="before")
    @classmethod
    def normalize_model(cls, data: Any) -> Any:
        if isinstance(data, dict):
            m = data.get("model")
            if isinstance(m, str):
                data["model"] = {"id": m, "speed": "standard"}
        return data


class AgentResponse(BaseModel):
    id: uuid.UUID
    type: str = "agent"
    name: str
    engine_kind: str
    model: Optional[ModelConfig] = None
    system: Optional[str] = None
    description: Optional[str] = None
    metadata: dict[str, str] = Field(default_factory=dict)
    env: dict[str, str] = Field(default_factory=dict)
    mcp_servers: list[McpServerConfig] = Field(default_factory=list)
    skills: list[PackedItem] = Field(default_factory=list)
    agents: list[PackedItem] = Field(default_factory=list)
    commands: list[PackedItem] = Field(default_factory=list)
    tools: list[AgentTool] = Field(default_factory=list)
    multiagent: Optional[dict[str, Any]] = None
    version: int
    environment_ref: Optional[str] = None
    secret_ref: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    archived_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("id")
    def serialize_id(self, v: uuid.UUID) -> str:
        return f"agent_{v}"


class AgentVersionResponse(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID
    version: int
    snapshot: dict[str, Any]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InjectConfig(BaseModel):
    name: str
    target: str
    tar_gz_b64: str


def extract_permission_mode(tools: list[dict]) -> str:
    for tool in tools:
        tool_type = tool.get("type", "")
        if tool_type in ("agent_toolset_20260401", "mcp_toolset"):
            dc = tool.get("default_config", {})
            pp = dc.get("permission_policy", {})
            if pp.get("type") == "always_ask":
                return "default"
    return "bypassPermissions"

File: sandbox-runner/crates/joysafeter-types/src/agent.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::{
    AgentId, CredentialGroupId, CredentialId, EnvironmentId, SandboxId, SessionId, TaskId,
};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EngineKind {
    Claude,
    Codex,
    Native,
}

impl std::fmt::Display for EngineKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Claude => write!(f, "claude"),
            Self::Codex => write!(f, "codex"),
            Self::Native => write!(f, "native"),
        }
    }
}

impl EngineKind {
    pub fn from_str_lossy(s: &str) -> Self {
        match s {
            "codex" => Self::Codex,
            "native" => Self::Native,
            _ => Self::Claude,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelConfig {
    pub id: String,
    #[serde(default = "default_model_speed")]
    pub speed: String,
}

fn default_model_speed() -> String {
    "standard".to_string()
}

impl ModelConfig {
    pub fn from_id(id: String) -> Self {
        Self {
            id,
            speed: default_model_speed(),
        }
    }
}

fn default_object_type() -> String {
    "agent".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AgentTool {
    #[serde(rename = "agent_toolset_20260401")]
    AgentToolset {
        #[serde(default)]
        default_config: Option<ToolDefaultConfig>,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        configs: Vec<ToolItemConfig>,
    },
    #[serde(rename = "mcp_toolset")]
    McpToolset {
        mcp_server_name: String,
        #[serde(default)]
        default_config: Option<ToolDefaultConfig>,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        configs: Vec<ToolItemConfig>,
    },
    #[serde(rename = "custom")]
    Custom {
        name: String,
        description: String,
        input_schema: serde_json::Value,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolItemConfig {
    pub name: String,
    #[serde(default = "default_true")]
    pub enabled: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub permission_policy: Option<PermissionPolicy>,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ToolDefaultConfig {
    #[serde(default)]
    pub permission_policy: PermissionPolicy,
    #[serde(default = "default_true", skip_serializing_if = "is_true")]
    pub enabled: bool,
}

fn is_true(v: &bool) -> bool {
    *v
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum PermissionPolicy {
    #[serde(rename = "always_allow")]
    #[default]
    AlwaysAllow,
    #[serde(rename = "always_ask")]
    AlwaysAsk,
}

impl PermissionPolicy {
    pub fn to_mode_str(&self) -> &'static str {
        match self {
            Self::AlwaysAllow => "bypassPermissions",
            Self::AlwaysAsk => "default",
        }
    }
}

pub fn extract_permission_mode(tools: &[AgentTool]) -> String {
    for tool in tools {
        match tool {
            AgentTool::AgentToolset {
                default_config,
                configs,
            } => {
                let base = default_config
                    .as_ref()
                    .map(|c| &c.permission_policy)
                    .unwrap_or(&PermissionPolicy::AlwaysAllow);
                if matches!(base, PermissionPolicy::AlwaysAsk) {
                    return "default".to_string();
                }
                if configs
                    .iter()
                    .any(|c| matches!(c.permission_policy, Some(PermissionPolicy::AlwaysAsk)))
                {
                    return "default".to_string();
                }
            }
            AgentTool::McpToolset {
                default_config,
                configs,
                ..
            } => {
                let base = default_config
                    .as_ref()
                    .map(|c| &c.permission_policy)
                    .unwrap_or(&PermissionPolicy::AlwaysAsk);
                if matches!(base, PermissionPolicy::AlwaysAsk) {
                    return "default".to_string();
                }
                if configs
                    .iter()
                    .any(|c| matches!(c.permission_policy, Some(PermissionPolicy::AlwaysAsk)))
                {
                    return "default".to_string();
                }
            }
            AgentTool::Custom { .. } => {}
        }
    }
    "bypassPermissions".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PackedItem {
    pub name: String,
    pub tar_gz_b64: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    pub id: AgentId,
    #[serde(rename = "type", default = "default_object_type")]
    pub object_type: String,
    pub name: String,
    pub engine_kind: EngineKind,
    pub model: ModelConfig,
    #[serde(rename = "system")]
    pub system_prompt: Option<String>,
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, String>,
    #[serde(default)]
    pub env: HashMap<String, String>,
    pub mcp_servers: Vec<McpServerConfig>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub skills: Vec<PackedItem>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub agents: Vec<PackedItem>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub commands: Vec<PackedItem>,
    pub tools: Vec<AgentTool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub multiagent: Option<serde_json::Value>,
    pub version: i32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub environment_id: Option<EnvironmentId>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_credential_id: Option<CredentialId>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub archived_at: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InjectConfig {
    pub name: String,
    pub target: String,
    pub tar_gz_b64: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum McpServerConfig {
    #[serde(rename = "streamable_http")]
    StreamableHttp { name: String, url: String },
    #[serde(rename = "sse")]
    Sse { name: String, url: String },
    #[serde(rename = "local_stdio")]
    LocalStdio {
        name: String,
        command: String,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        args: Vec<String>,
        #[serde(default, skip_serializing_if = "HashMap::is_empty")]
        env: HashMap<String, String>,
    },
}

impl McpServerConfig {
    pub fn name(&self) -> &str {
        match self {
            Self::StreamableHttp { name, .. }
            | Self::Sse { name, .. }
            | Self::LocalStdio { name, .. } => name,
        }
    }
}

#[cfg(test)]
mod mcp_server_config_tests {
    use std::collections::HashMap;

    use super::McpServerConfig;

    #[test]
    fn accepts_only_canonical_mcp_transports() {
        let parsed: McpServerConfig = serde_json::from_value(serde_json::json!({
            "type": "streamable_http",
            "name": "remote",
            "url": "https://example.com/mcp"
        }))
        .unwrap();
        assert!(matches!(parsed, McpServerConfig::StreamableHttp { .. }));
        let sse: McpServerConfig = serde_json::from_value(serde_json::json!({
            "type": "sse", "name": "events", "url": "https://example.com/sse"
        }))
        .unwrap();
        assert!(matches!(sse, McpServerConfig::Sse { .. }));
        let local: McpServerConfig = serde_json::from_value(serde_json::json!({
            "type": "local_stdio", "name": "local", "command": "node",
            "args": ["server.js"], "env": {"MODE": "safe"}
        }))
        .unwrap();
        assert_eq!(
            local,
            McpServerConfig::LocalStdio {
                name: "local".to_string(),
                command: "node".to_string(),
                args: vec!["server.js".to_string()],
                env: HashMap::from([("MODE".to_string(), "safe".to_string())]),
            }
        );

        for transport in ["url", "http", "streamable-http", "stdio"] {
            assert!(
                serde_json::from_value::<McpServerConfig>(serde_json::json!({
                    "type": transport,
                    "name": "legacy",
                    "url": "https://example.com/mcp"
                }))
                .is_err()
            );
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatSession {
    pub id: SessionId,
    pub agent_id: AgentId,
    pub title: Option<String>,
    pub status: String,
    pub environment_id: Option<EnvironmentId>,
    pub last_harness_session_id: Option<String>,
    pub last_work_dir: Option<String>,
    pub last_sandbox_id: Option<SandboxId>,
    #[serde(default)]
    pub agent_version: Option<i32>,
    #[serde(default)]
    pub agent_snapshot: Option<serde_json::Value>,
    #[serde(default)]
    pub credential_group_ids: Vec<CredentialGroupId>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SandboxRecord {
    pub id: SandboxId,
    pub external_id: String,
    pub provider: String,
    pub status: String,
    pub config: serde_json::Value,
    pub chat_session_id: Option<SessionId>,
    pub image: String,
    pub last_task_id: Option<TaskId>,
    pub last_used_at: chrono::DateTime<chrono::Utc>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub destroyed_at: Option<chrono::DateTime<chrono::Utc>>,
    pub workspace_path: Option<String>,
}
